File: BlinkyHeadless/CPP/CppBlinkyStartup/StartupTask/BackgroundTask.cpp
 /********************************************************
*                                                        *
*   © Copyright (C) Microsoft. All rights reserved.      *
*                                                        *
*********************************************************/


#include "pch.h"
#include "BackgroundTask.h"

using namespace Windows::Foundation;
using namespace Windows::Devices::Enumeration;
using namespace Windows::Devices::Gpio;
using namespace concurrency;

using namespace StartupTask;
using namespace Platform;

BackgroundTask::BackgroundTask()
{
}

void BackgroundTask::Run(IBackgroundTaskInstance^ taskInstance)
{
    Deferral = taskInstance->GetDeferral();

    TimerElapsedHandler ^handler = ref new TimerElapsedHandler(
        [this](ThreadPoolTimer ^timer) 
        {
            if (pin != nullptr)
            {
                if (LEDStatus == 0)
                {
                    LEDStatus = 1;
					pin->Write(GpioPinValue::High);
                }
                else
                {
                    LEDStatus = 0;
					pin->Write(GpioPinValue::Low);
                }
            }
        }
    );
    TimeSpan interval;
    interval.Duration = 500 * 1000 * 10;
    Timer = ThreadPoolTimer::CreatePeriodicTimer(handler, interval);

    InitGpio();
}

void BackgroundTask::InitGpio()
{
	create_task(GpioController::GetDefaultAsync()).then([this](task<GpioController ^> controllerOp) {
		auto gpio = controllerOp.get();
		pin = gpio->OpenPin(LED_PIN);

		if (pin != nullptr)
		{
			pin->Write(GpioPinValue::High);
			pin->SetDriveMode(GpioPinDriveMode::Output);
		}
	});
}

File: BlinkyHeadless/CS/CsBlinkyStartup/StartupTask/BackgroundTask.cs
// © Copyright (C) Microsoft. All rights reserved.


using System;
using Windows.ApplicationModel.Background;
using Windows.Devices.Gpio;
using Windows.System.Threading;

namespace StartupTask
{
    public sealed class BackgroundTask : IBackgroundTask
    {
        BackgroundTaskDeferral _deferral;
        public void Run(IBackgroundTaskInstance taskInstance)
        {
            _deferral = taskInstance.GetDeferral();

            timer = ThreadPoolTimer.CreatePeriodicTimer(Timer_Tick, TimeSpan.FromMilliseconds(500));
            InitGPIO();
        }

        private async void InitGPIO()
        {
            var gpio = await GpioController.GetDefaultAsync();
            pin = gpio.OpenPin(LED_PIN);

            if (pin == null)
            {
                return;
            }

            pin.Write(GpioPinValue.High);
            pin.SetDriveMode(GpioPinDriveMode.Output);
        }

        private void Timer_Tick(ThreadPoolTimer timer)
        {
            if (pin != null)
            {
                if (LEDStatus == 0)
                {
                    LEDStatus = 1;
                    pin.Write(GpioPinValue.High);
                }
                else
                {
                    LEDStatus = 0;
                    pin.Write(GpioPinValue.Low);
                }
            }
        }

        private int LEDStatus = 0;
        private const int LED_PIN = 0;
        private GpioPin pin;
        private ThreadPoolTimer timer;
    }
}
